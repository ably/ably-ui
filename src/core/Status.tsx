import React from "react";
import useSWR from "swr";
import cn from "./utils/cn";
import Icon from "./Icon";

type StatusProps = {
  statusUrl: string;
  additionalCSS?: string;
  refreshInterval?: number;
  showDescription?: boolean;
};

export const statusTypes = [
  "none",
  "operational",
  "minor",
  "major",
  "critical",
  "unknown",
] as const;

export type StatusType = (typeof statusTypes)[number];

export const StatusUrl =
  "https://ntqy1wz94gjv.statuspage.io/api/v2/status.json";

// Our SWR fetcher function
const fetcher = (url: string) => fetch(url).then((res) => res.json());

const indicatorClass = (indicator?: StatusType) => {
  switch (indicator) {
    case "none":
    case "operational":
      return "bg-gui-success-green";
    case "minor":
      return "bg-yellow-500";
    case "major":
      return "bg-orange-500";
    case "critical":
      return "bg-gui-error-red";
    default:
      return "bg-neutral-500";
  }
};

export const StatusIcon = ({
  statusUrl,
  refreshInterval = 1000 * 60,
}: StatusProps) => {
  const { data, error, isLoading } = useSWR(statusUrl, fetcher, {
    refreshInterval,
  });

  return (
    <span
      className={cn(
        "inline-flex h-2 aspect-square rounded-full",
        indicatorClass(data?.status?.indicator),
        { "animate-pulse": isLoading || error },
      )}
    ></span>
  );
};

const Status = ({
  statusUrl = StatusUrl,
  additionalCSS,
  refreshInterval = 1000 * 60,
  showDescription = false,
}: StatusProps) => {
  const { data } = useSWR(statusUrl, fetcher, {
    refreshInterval,
  });

  return (
    <a
      href="https://status.ably.com"
      className={cn(
        "inline-flex group/status items-center gap-2",
        additionalCSS,
      )}
      target="_blank"
      rel="noreferrer"
    >
      <StatusIcon
        statusUrl={statusUrl}
        refreshInterval={refreshInterval ?? 1000 * 60}
      />
      {showDescription && data?.status?.description && (
        <div className="flex gap-2 ui-text-label4 font-medium text-neutral-900 group-hover/status:text-ably-primary dark:text-neutral-400 transition-colors">
          <span>
            {data.status.description.charAt(0).toUpperCase() +
              data.status.description.slice(1).toLowerCase()}
          </span>
          <Icon name="icon-gui-arrow-top-right-on-square-outline" size="16px" />
        </div>
      )}
    </a>
  );
};

export default Status;
