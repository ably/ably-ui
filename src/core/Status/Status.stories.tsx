import React from "react";
import { delay, http, HttpResponse } from "msw";
import Status from "../Status";

const statusUrl = "https://ntqy1wz94gjv.statuspage.io/api/v2/status.json";

export default {
  title: "JS Components/Status",
  component: Status,
  args: {
    statusUrl,
  },
  tags: ["!autodocs"],
};

export const Loading = {
  parameters: {
    msw: {
      handlers: {
        statusNone: http.get(statusUrl, async () => {
          await delay("infinite");

          return HttpResponse.json({
            status: {
              indicator: "will never arrive",
            },
          });
        }),
      },
    },
  },
  render: () => <Status statusUrl={statusUrl} />,
};

const mockParametersWithStatus = (indicator) => {
  return {
    msw: {
      handlers: {
        status: http.get(statusUrl, async () => {
          await delay();

          return HttpResponse.json({
            status: {
              indicator: indicator,
            },
          });
        }),
      },
    },
  };
};

export const None = {
  parameters: mockParametersWithStatus("none"),
  render: () => <Status statusUrl={statusUrl} />,
};

export const Operational = {
  parameters: mockParametersWithStatus("operational"),
  render: () => <Status statusUrl={statusUrl} />,
};

export const Minor = {
  parameters: mockParametersWithStatus("minor"),
  render: () => <Status statusUrl={statusUrl} />,
};

export const Major = {
  parameters: mockParametersWithStatus("major"),
  render: () => <Status statusUrl={statusUrl} />,
};

export const Critical = {
  parameters: mockParametersWithStatus("critical"),
  render: () => <Status statusUrl={statusUrl} />,
};

export const Unknown = {
  parameters: mockParametersWithStatus("unknown"),
  render: () => <Status statusUrl={statusUrl} />,
};
