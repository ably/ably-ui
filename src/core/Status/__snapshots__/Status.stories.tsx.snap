// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS Components/Status Critical smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;

exports[`JS Components/Status Loading smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;

exports[`JS Components/Status Major smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;

exports[`JS Components/Status Minor smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;

exports[`JS Components/Status None smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;

exports[`JS Components/Status Operational smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;

exports[`JS Components/Status Unknown smoke-test 1`] = `
<a href="https://status.ably.com"
   class="inline-block undefined"
   target="_blank"
   rel="noreferrer"
>
  <span class="flex items-center h-[1.5rem] p-[0.25rem]">
    <span class="w-[1rem] h-[1rem] leading-[1rem] rounded-full animate-pulse bg-neutral-500">
    </span>
  </span>
</a>
`;
